### PR TITLE
Typing

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

**Comment by benma** [comment]
How to have CI run on this PR? Would be good to see if it passes.

File: hwilib/serializations.py [hunk 1]
```diff
@@ -792,8 +813,10 @@ def deserialize(self, psbt):
             input.deserialize(f)
             self.inputs.append(input)
 
-            if input.non_witness_utxo and input.non_witness_utxo.rehash() and input.non_witness_utxo.sha256 != txin.prevout.sha256:
```
Comment (benma) [inline hwilib/serializations.py:795]: Typing not only helps for code clarity, but also helps to catch issues like this (`prevout.sha256` does not exist, and `rehash()` never returns anything).

File: hwilib/serializations.py [hunk 2]
```diff
@@ -25,22 +24,43 @@
...
+    Dict,
+    List,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    TypeVar,
+Callable,
```
Comment (achow101) [inline hwilib/serializations.py:37]: nit: indentation

Comment (benma) [reply hwilib/serializations.py:37]: Fixed.

**Comment by benma** [comment]
> Not sure why travis isn't running..

Maybe Travis flagged this org as suspicious, happened to me once. If so, their support will clear it up quickly.

**Comment by promag** [comment]
@benma don't think so, travis runs for https://github.com/bitcoin-core/gui.

**Comment by achow101** [comment]
> Maybe Travis flagged this org as suspicious

Doesn't seem like it as I'm able to manually trigger a build of the master branch.

**Comment by achow101** [comment]
I pushed the earlier version of this to a separate branch in this repo to get a travis run: https://travis-ci.org/github/bitcoin-core/HWI/builds/710516659. There appear to be a few linter errors.

**Comment by benma** [comment]
> I pushed the earlier version of this to a separate branch in this repo to get a travis run: https://travis-ci.org/github/bitcoin-core/HWI/builds/710516659. There appear to be a few linter errors.

Thanks. Updated mypy locally and it indeed found more relevant issues. Great.

I force pushed a fix, please try.

**Comment by benma** [comment]
https://travis-ci.org/github/bitcoin-core/HWI/jobs/710561930

Strange that this error is caught now, seems unrelated to my PR.

Anyway, attempted a fix :crossed_fingers:

File: hwilib/serializations.py [hunk 3]
```diff
@@ -25,22 +24,47 @@
...
+    Tuple,
+    TypeVar,
+    Callable,
+)
+
+from typing_extensions import Protocol
+
+class Readable(Protocol):
+    def read(self, n: int = -1) -> bytes:
+        ...
```
[the remark this answers is not in the document]
Comment (benma) [reply hwilib/serializations.py:43]: NotImplementError would be misleading, as it is not supposed to be inherited and implemented.

`...` and `pass` are equivalent as is a matter of style only. I usually put it in type definitions like here, to visually distinguish it from a function that is actually meant to be called at runtime.

Lmk which way you prefer, I don't have a strong opinion.

**Comment by achow101** [comment]
ACK 6ad1f8924fa0cb734902747ff755c2cfcd90af0b

**Pull request merged**